Framework/PythonInterface/core/inc/MantidPythonInterface/core/Converters/PyNativeTypeExtractor.h

```diff
@@ -0,0 +1,135 @@
+}
+
+namespace Mantid::PythonInterface {
+
+struct PyNativeTypeExtractor {
+  using PythonOutputT =
+      boost::make_recursive_variant<bool, long, double, std::string, std::vector<boost::recursive_variant_>>::type;
+
+  static PythonOutputT convert(const boost::python::object &obj) {
+    using namespace boost::python;
+    PyObject *rawptr = obj.ptr();
+    PythonOutputT out;
+
+    if (PyList_Check(rawptr)) {
+      out = handleList(obj);
+    } else if (PyBool_Check(rawptr)) {
+      out = extract<bool>(obj);
+    } else if (PyFloat_Check(rawptr)) {
+      out = extract<double>(obj);
+    } else if (PyLong_Check(rawptr)) {
+      out = extract<long>(obj);
+    } else if (PyUnicode_Check(rawptr)) {
+      out = extract<std::string>(obj);
+    } else {
+      throw std::invalid_argument("Unrecognised Python type");
+    }
+    return out;
+  }
+
+private:
+  static PythonOutputT handleList(const boost::python::object &obj) {
+    auto rawptr = obj.ptr();
+    auto n = PyList_Size(rawptr);
+    auto vec = std::vector<PythonOutputT>();
+    vec.reserve(n);
+    for (Py_ssize_t i = 0; i < n; ++i) {
+      vec.emplace_back(convert(obj[i]));
+    }
+    return vec;
+  }
+};
+
+class IPyTypeVisitor : public boost::static_visitor<> {
+public:
+  /**
+   * Dynamically dispatches to overloaded operator depending on the underlying type.
+   * This also handles cases such as std::vector<T>, or nested lists, which will be flattened
+   * by invoking the operator each time for each elem.
+   * This assumes all element in a list matches the first element type found in that list,
+   * where this is not true, the element will be cast to the first element type.
+   *
+   * Note: you will need to include
+   * using Mantid::PythonInterface::IPyTypeVisitor::operator();
+   */
+
+  virtual ~IPyTypeVisitor() = default;
+  virtual void operator()(bool value) const = 0;
+  virtual void operator()(long value) const = 0;
+  virtual void operator()(double value) const = 0;
+  virtual void operator()(std::string) const = 0;
+
+  virtual void operator()(std::vector<bool>) const = 0;
+  virtual void operator()(std::vector<long>) const = 0;
+  virtual void operator()(std::vector<double>) const = 0;
+  virtual void operator()(std::vector<std::string>) const = 0;
+
+  void operator()(std::vector<PyNativeTypeExtractor::PythonOutputT> const &values) const {
+    if (values.size() == 0)
+      return;
+    const auto &elemType = values[0].type();
+
+    // We must manually dispatch for container types, as boost will try
+    // to recurse down to scalar values.
+    if (elemType == typeid(bool)) {
+      applyVectorProp<bool>(values);
+    } else if (elemType == typeid(double)) {
+      applyVectorProp<double>(values);
+    } else if (elemType == typeid(long)) {
+      applyVectorProp<long>(values);
+    } else if (elemType == typeid(std::string)) {
+      applyVectorProp<std::string>(values);
+    } else {
+      // Recurse down
+      for (const auto &val : values) {
```

Comment:
> Also I would (maybe there isn't a cast for it) never be able to set a `std::vector<std::vector<Type>>` property right?

Hmm I thought our intent was to support vectors of vectors but I think you're right, it will set the same property twice (or fail). @DavidFair do you agree?

Reply:
We should handle vectors of vectors, but it does the former behaviour - calls the `std::vector<T>` operator two times

If we want to detect this we should add a check to `Algorithm.cpp` as there may be cases where this is valid.
If, in the future we want two layer deep vectors we will need to extend the manual `typeID` checks to stop boost recursing down to scalar values